tests/core/service/datafeed/abstract_datafeed_loop_test.py

```diff
@@ -0,0 +1,480 @@
+
+    await handle_events_and_wait_for_completion(df_loop, [event])
+
+    listener.is_accepting_event.assert_called_with(event, BOT_USER)
+    listener.on_connection_accepted.assert_called_with(initiator, payload.connection_accepted)
+
+
+@pytest.mark.asyncio
+async def test_handle_message_suppressed(df_loop, listener, initiator):
+    payload = V4Payload(message_suppressed=V4MessageSuppressed())
+    event = V4Event(type=RealTimeEvent.MESSAGESUPPRESSED.name, payload=payload, initiator=initiator)
+
+    await handle_events_and_wait_for_completion(df_loop, [event])
+
+    listener.is_accepting_event.assert_called_with(event, BOT_USER)
+    listener.on_message_suppressed.assert_called_with(initiator, payload.message_suppressed)
+
+
+@pytest.mark.asyncio
+async def test_handle_symphony_element(df_loop, listener, initiator):
+    payload = V4Payload(symphony_elements_action=V4SymphonyElementsAction())
+    event = V4Event(type=RealTimeEvent.SYMPHONYELEMENTSACTION.name, payload=payload, initiator=initiator)
+
+    await handle_events_and_wait_for_completion(df_loop, [event])
+
+    listener.is_accepting_event.assert_called_with(event, BOT_USER)
+    listener.on_symphony_elements_action.assert_called_once_with(initiator, payload.symphony_elements_action)
+
+
+@pytest.mark.asyncio
+async def test_listener_concurrency(bare_df_loop, read_df_side_effect):
+    class QueueListener(RealTimeEventListener):
+        def __init__(self, queue, other_queue):
+            self.queue = queue
+            self.other_queue = other_queue
+            self.first = True
+
+        async def on_message_sent(self, initiator: V4Initiator, event: V4MessageSent):
+            if self.first:
+                await self.queue.put("message")
+                await self.other_queue.get()
+                await bare_df_loop.stop()
+            self.first = False
+
+    bare_df_loop.read_datafeed.side_effect = read_df_side_effect
+
+    queue_one = asyncio.Queue()
+    queue_two = asyncio.Queue()
+
+    bare_df_loop.subscribe(QueueListener(queue_one, queue_two))
+    bare_df_loop.subscribe(QueueListener(queue_two, queue_one))
+
+    await bare_df_loop.start()  # test it finishes without deadlock
+
+
+@pytest.mark.asyncio
+async def test_events_concurrency(bare_df_loop, read_df_side_effect):
+    class QueueListener(RealTimeEventListener):
+        def __init__(self):
+            self.queue = asyncio.Queue()
+            self.count = 0
+
+        async def on_message_sent(self, initiator: V4Initiator, event: V4MessageSent):
+            self.count += 1
+            if self.count == 1:
+                await self.queue.get()
+                await bare_df_loop.stop()
+            elif self.count == 2:
+                await self.queue.put("message")
+
+    bare_df_loop.read_datafeed.side_effect = read_df_side_effect
+    bare_df_loop.subscribe(QueueListener())
+
+    await bare_df_loop.start()  # test no deadlock
```

Comment:
is there a way to set a maximum timeout for a python test so that in case of deadlock we don't block the CI build?

Reply:
not out of the box maybe (https://pypi.org/project/pytest-timeout/)
not sure it is worth adding an extra library

Reply:
maybe https://stackoverflow.com/questions/59073731/set-default-timeout-on-github-action-pipeline is simpler